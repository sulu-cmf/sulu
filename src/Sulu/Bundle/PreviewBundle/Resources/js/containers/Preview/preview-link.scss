@import 'sulu-admin-bundle/containers/Application/colors.scss';

.container {
    padding: 20px;
}

.input-container {
    display: flex;
}

.label {
    color: $silverChalice;
    display: block;
    height: 20px;
}

.input {
    background-color: $blueZodiac;
    border: 1px solid $doveGray;
    color: $white;
    display: block;
    border-radius: 3px;
    height: 30px;
    width: 240px;
    padding: 5px;
}

.copy-button {
    min-width: 90px;
    margin-left: 10px;
    padding-left: 10px;
    padding-right: 10px;
}

.revoke {
    margin-top: 20px;
}

.revoke-button {
    color: $white;
    font-size: 14px;
    text-decoration: underline;
    background: transparent;
    border: none;
    cursor: pointer;

    &:hover {
        color: $shakespeare;
    }
}
