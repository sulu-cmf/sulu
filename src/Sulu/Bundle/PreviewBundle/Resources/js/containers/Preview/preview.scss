@import '../../../../../AdminBundle/Resources/js/containers/Application/colors.scss';

$previewBackgroundColor: $white;
$toolbarHeight: 60px;

.container {
    height: 100%;
    background-color: $previewBackgroundColor;
}

.iframe {
    width: 100%;
    height: 100%;
    border: 0;
}

.iframe-container {
    width: 100%;
    height: calc(100% - $toolbarHeight);
}

.loader-container {
    width: 100%;
    height: calc(100% - $toolbarHeight);
    padding-top: 100px;
}
