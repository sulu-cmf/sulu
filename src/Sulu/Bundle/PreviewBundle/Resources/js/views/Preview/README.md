The `Preview` sidebar-view is registered with the key `sulu_preview.preview`.
It shows the rendered web-page of the resource represented by the given resource-store.

To initialize the `Preview` for your component use following code-snippet:

```javascript static
const PageWithSidebar = withSidebar(Page, function() {
    const {
        router,
        resourceStore,
    } = this.props;

    return {
        view: 'sulu_preview.preview',
        sizes: ['medium', 'large'],
        props: {
            router: router,
            resourceStore: resourceStore,
        },
    };
});
``` 
