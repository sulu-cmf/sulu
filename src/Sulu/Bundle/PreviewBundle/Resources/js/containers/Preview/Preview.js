// @flow
import React from 'react';
import ReactDOM from 'react-dom';
import {action, autorun, observable, toJS} from 'mobx';
import {observer} from 'mobx-react';
import debounce from 'debounce';
import {Router} from 'sulu-admin-bundle/services';
import {FormStore, sidebarStore} from 'sulu-admin-bundle/containers';
import {Loader, Toolbar} from 'sulu-admin-bundle/components';
import previewStyles from './preview.scss';
import PreviewStore from './stores/PreviewStore';
import type {PreviewMode} from './types';

type Props = {|
    formStore: FormStore,
    router: Router,
|};

@observer
export default class Preview extends React.Component<Props> {
    static debounceDelay: number = 250;
    static mode: PreviewMode = 'auto';

    @observable iframeRef: ?HTMLIFrameElement;
    @observable started: boolean = false;

    previewStore: PreviewStore;

    typeDisposer: () => void;
    dataDisposer: () => void;

    constructor(props: Props) {
        super(props);

        const {
            formStore,
            router: {
                attributes: {
                    locale,
                    webspace,
                },
            },
        } = this.props;

        this.previewStore = new PreviewStore(formStore.resourceKey, formStore.id, locale, webspace);
    }

    componentDidMount() {
        if (Preview.mode === 'auto') {
            this.startPreview();
        }
    }

    @action setStarted = (started: boolean) => {
        this.started = started;
    };

    startPreview = () => {
        const {
            formStore,
        } = this.props;

        this.previewStore.start();

        this.dataDisposer = autorun(() => {
            if (this.previewStore.starting || formStore.loading || !this.iframeRef) {
                return;
            }

            this.updatePreview(toJS(formStore.data));
        });

        this.typeDisposer = autorun(() => {
            if (this.previewStore.starting || formStore.loading || !this.iframeRef) {
                return;
            }

            this.previewStore.updateContext(toJS(formStore.type)).then(this.setContent);
        });

        this.setStarted(true);
    };

    updatePreview = debounce((data: Object) => {
        this.previewStore.update(data).then(this.setContent);
    }, Preview.debounceDelay);

    setContent = (content: string) => {
        const document = this.getPreviewDocument();
        if (!document) {
            return;
        }

        document.open();
        document.write(content);
        document.close();
    };

    componentWillUnmount() {
        if (this.typeDisposer) {
            this.typeDisposer();
        }

        if (this.dataDisposer) {
            this.dataDisposer();
        }

        if (!this.started) {
            return;
        }

        this.previewStore.stop();
    }

    getPreviewDocument = (): ?Document => {
        // eslint-disable-next-line react/no-find-dom-node
        const iframe = ReactDOM.findDOMNode(this.iframeRef);
        if (!(iframe instanceof HTMLIFrameElement)) {
            return;
        }

        return iframe.contentDocument;
    };

    @action setIframe = (iframeRef: ?Object) => {
        this.iframeRef = iframeRef;
    };

    handleToggleSidebarClick = () => {
        if (sidebarStore.size === 'medium') {
            return sidebarStore.setSize('large');
        }

        sidebarStore.setSize('medium');
    };

    handleStartClick = () => {
        this.startPreview();
    };

    renderToolbar() {
        return (
            <Toolbar skin="dark">
                <Toolbar.Controls>
                    <Toolbar.Button
                        icon={sidebarStore.size === 'medium' ? 'su-arrow-left' : 'su-arrow-right'}
                        onClick={this.handleToggleSidebarClick}
                    />
                </Toolbar.Controls>
            </Toolbar>
        );
    }

    render() {
        if (!this.started) {
            return <button onClick={this.handleStartClick}>Start</button>;
        }

        if (this.previewStore.starting) {
            return (
                <div className={previewStyles.container}>
                    <div className={previewStyles.loaderContainer}>
                        <Loader />
                    </div>

                    {this.renderToolbar()}
                </div>
            );
        }

        return (
            <div className={previewStyles.container}>
                <div className={previewStyles.iframeContainer}>
                    <iframe
                        className={previewStyles.iframe}
                        ref={this.setIframe}
                        src={this.previewStore.renderRoute}
                    />
                </div>

                {this.renderToolbar()}
            </div>
        );
    }
}
