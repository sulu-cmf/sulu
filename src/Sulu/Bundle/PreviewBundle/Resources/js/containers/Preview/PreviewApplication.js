// @flow
import React from 'react';
import {action, computed, observable} from 'mobx';
import {observer} from 'mobx-react';
import Preview from './Preview';
import 'sulu-admin-bundle/containers/Application/global.scss';
import previewStyles from './preview.scss';
import {webspaceStore} from 'sulu-page-bundle/stores';
import {buildQueryString, transformDateForUrl} from 'sulu-admin-bundle/utils';
import debounce from 'debounce';

const Props = {};

@observer
class PreviewApplication extends React.Component<Props> {
    static debounceDelay: number = 250;

    @observable webspaceKey: string;
    @observable webspaceOptions: Array<Object> = [];

    @observable segmentKey: string;
    @computed get segments() {
        if (!this.webspaceKey) {
            return [];
        }

        return webspaceStore.getWebspace(this.webspaceKey).segments;
    }

    @observable dateTime: Date;

    @observable iframeRef: ?HTMLIFrameElement;
    @observable reloadCounter: number = 0;

    @computed get renderRoute() {
        return document.location.href + '/render' + buildQueryString({
            webspaceKey: this.webspaceKey,
            segmentKey: this.segmentKey,
            dateTime: this.dateTime && transformDateForUrl(this.dateTime),
        });
    }

    constructor(props: Props) {
        super(props);

        this.webspaceOptions = webspaceStore.allWebspaces.map((webspace): Object => ({
            label: webspace.name,
            value: webspace.key,
        }));
        this.webspaceKey = this.webspaceOptions[0].value;

        this.dateTime = new Date();
    }

    @action handleSegmentChange = (segmentKey: ?string) => {
        this.segmentKey = segmentKey;
    };

    @action handleDateTimeChange = debounce((value: ?Date) => {
        this.dateTime = value || new Date();
    }, PreviewApplication.debounceDelay);

    @action handleWebspaceChange = (webspaceKey: string) => {
       this.webspaceKey = webspaceKey;
    };

    @action handleRefreshClick = () => {
        // We can not reload the iframe here as safari and firefox
        // resets the location.href to another url on previewDocument.open
        // so instead of this we rerender the whole iframe.
        ++this.reloadCounter;
    };

    @action setIframe = (iframeRef: ?Object) => {
        this.iframeRef = iframeRef;
    };

    render() {
        return (
            <div className={previewStyles.applicationContainer}>
                <Preview
                    webspace={this.webspaceKey}
                    webspaceOptions={this.webspaceOptions}
                    segments={this.segments}
                    segment={this.segmentKey}
                    targetGroup={null}
                    targetGroupOptions={[]}
                    size={null}
                    dateTime={this.dateTime}
                    renderRoute={this.renderRoute}
                    onPreviewWindowClick={null}
                    onSegmentChange={this.handleSegmentChange}
                    onTargetGroupChange={null}
                    onDateTimeChange={this.handleDateTimeChange}
                    onToggleSidebarClick={null}
                    onWebspaceChange={null /* TODO for e.g. articles */}
                    onRefreshClick={this.handleRefreshClick}
                >
                    <iframe
                        className={previewStyles.iframe}
                        key={this.reloadCounter}
                        ref={this.setIframe}
                        src={this.renderRoute}
                    />
                </Preview>
            </div>
        );
    }
}

export default PreviewApplication;
