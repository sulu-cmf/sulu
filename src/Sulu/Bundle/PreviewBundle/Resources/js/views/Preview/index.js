// @flow
import Preview from './Preview';
import PreviewStore from './stores/PreviewStore';

export {PreviewStore};
export default Preview;
