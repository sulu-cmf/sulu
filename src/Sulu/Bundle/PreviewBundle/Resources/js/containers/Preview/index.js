// @flow
import Preview from './Preview';
import PreviewApplication from './PreviewApplication';
import PreviewSidebar from './PreviewSidebar';
import PreviewStore from './stores/PreviewStore';

export {PreviewApplication, PreviewSidebar, PreviewStore};
export default Preview;
