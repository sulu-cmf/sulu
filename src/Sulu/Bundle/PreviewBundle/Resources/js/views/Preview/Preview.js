// @flow
import React from 'react';
import ReactDOM from 'react-dom';
import {action, autorun, observable, toJS} from 'mobx';
import {observer} from 'mobx-react';
import debounce from 'debounce';
import Router from 'sulu-admin-bundle/services/Router';
import {sidebarStore} from 'sulu-admin-bundle/containers/Sidebar';
import {FormStore} from 'sulu-admin-bundle/containers/Form';
import ResourceStore from 'sulu-admin-bundle/stores/ResourceStore';
import Requester from 'sulu-admin-bundle/services/Requester';
import Toolbar from 'sulu-admin-bundle/components/Toolbar';
import Loader from 'sulu-admin-bundle/components/Loader';
import previewStyles from './preview.scss';
import previewConfigStore from './stores/PreviewConfigStore';

type Props = {
    resourceStore: ResourceStore,
    formStore: FormStore,
    router: Router,
};

@observer
export default class Preview extends React.Component<Props> {
    @observable iframeRef: ?Object;
    @observable token: ?string;
    @observable started: boolean = false;

    typeDisposer: () => void;
    dataDisposer: () => void;

    componentDidMount() {
        if (previewConfigStore.mode === 'auto') {
            this.startPreview();
        }
    }

    @action startPreview = () => {
        const {
            resourceStore,
            formStore,
            router: {
                attributes: {
                    locale,
                },
            },
        } = this.props;

        const route = previewConfigStore.generateRoute('start', {
            provider: resourceStore.resourceKey,
            locale: locale,
            id: resourceStore.id,
        });
        if (!route) {
            return;
        }

        Requester.get(route).then((response) => {
            this.setToken(response.token);
        });

        this.typeDisposer = autorun(() => {
            if (formStore.loading || !this.iframeRef) {
                return;
            }

            // $FlowFixMe
            this.updateContext(formStore.type.get());
        });

        this.dataDisposer = autorun(() => {
            if (resourceStore.loading || !this.iframeRef) {
                return;
            }

            this.updatePreview(toJS(resourceStore.data));
        });

        this.started = true;
    };

    updatePreview = debounce((data) => {
        const {
            router: {
                attributes: {
                    locale,
                    webspace,
                },
            },
        } = this.props;

        const route = previewConfigStore.generateRoute('update', {
            locale: locale,
            webspace: webspace,
            token: this.token,
        });
        if (!route) {
            return;
        }

        Requester.post(route, {data: data}).then((response) => {
            const document = this.getPreviewDocument();
            if (!document) {
                return;
            }

            document.open();
            document.write(response.content);
            document.close();
        });
    }, previewConfigStore.debounceDelay);

    updateContext = debounce((type) => {
        const {
            router: {
                attributes: {
                    webspace,
                },
            },
        } = this.props;

        const route = previewConfigStore.generateRoute('update-context', {
            webspace: webspace,
            token: this.token,
        });
        if (!route) {
            return;
        }

        Requester.post(route, {context: {template: type}}).then((response) => {
            const document = this.getPreviewDocument();
            if (!document) {
                return;
            }

            document.open();
            document.write(response.content);
            document.close();
        });
    }, previewConfigStore.debounceDelay);

    componentWillUnmount() {
        if (this.typeDisposer) {
            this.typeDisposer();
        }

        if (this.dataDisposer) {
            this.dataDisposer();
        }

        if (!this.started) {
            return;
        }

        const route = previewConfigStore.generateRoute('stop', {token: this.token});
        if (!route) {
            return;
        }

        Requester.get(route);
    }

    getPreviewDocument = (): ?Document => {
        // eslint-disable-next-line
        const iframe = ReactDOM.findDOMNode(this.iframeRef);
        if (!(iframe instanceof HTMLIFrameElement)) {
            return;
        }

        return iframe.contentDocument;
    };

    @action setToken = (token: string) => {
        this.token = token;
    };

    @action setIframe = (iframeRef: ?Object) => {
        this.iframeRef = iframeRef;
    };

    handleToggleSidebarClick = () => {
        if (sidebarStore.size === 'medium') {
            return sidebarStore.setSize('large');
        }

        sidebarStore.setSize('medium');
    };

    handleRefreshClick = () => {
        const document = this.getPreviewDocument();
        if (!document) {
            return;
        }

        document.location.reload();
    };

    handleStartClick = () => {
        this.startPreview();
    };

    renderToolbar() {
        return (
            <Toolbar skin="dark">
                <Toolbar.Controls>
                    <Toolbar.Button
                        icon={sidebarStore.size === 'medium' ? 'su-arrow-left' : 'su-arrow-right'}
                        onClick={this.handleToggleSidebarClick}
                    />
                    <Toolbar.Button
                        icon="fa-refresh"
                        onClick={this.handleRefreshClick}
                    />
                </Toolbar.Controls>
            </Toolbar>
        );
    }

    render() {
        if (!this.started) {
            return <button onClick={this.handleStartClick}>Start</button>;
        }

        if (!this.token) {
            return (
                <div className={previewStyles.container}>
                    <div className={previewStyles.loaderContainer}>
                        <Loader />
                    </div>

                    {this.renderToolbar()}
                </div>
            );
        }

        const {
            router: {
                attributes: {
                    locale,
                    webspace,
                },
            },
        } = this.props;

        const url = previewConfigStore.generateRoute('render', {
            webspace: webspace,
            locale: locale,
            token: this.token,
        });

        return (
            <div className={previewStyles.container}>
                <div className={previewStyles.iframeContainer}>
                    <iframe
                        className={previewStyles.iframe}
                        ref={this.setIframe}
                        src={url}
                    />
                </div>

                {this.renderToolbar()}
            </div>
        );
    }
}
