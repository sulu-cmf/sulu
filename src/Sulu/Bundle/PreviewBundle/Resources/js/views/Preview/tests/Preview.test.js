// @flow
import React from 'react';
import {observable} from 'mobx';
import {mount, render, shallow} from 'enzyme';
import Requester from 'sulu-admin-bundle/services/Requester';
import ResourceStore from 'sulu-admin-bundle/stores/ResourceStore';
import Router from 'sulu-admin-bundle/services/Router';
import Preview from '../Preview';
import type {PreviewRouteName} from '../stores/PreviewConfigStore';
import previewConfigStore from '../stores/PreviewConfigStore';

jest.mock('debounce', () => jest.fn((value) => value));

jest.mock('../stores/PreviewConfigStore', () => ({
    setConfig(config) {
        this.mode = config.mode;
    },
    mode: 'auto',
    generateRoute: (name: PreviewRouteName) => {
        return '/' + name;
    },
}));

jest.mock('sulu-admin-bundle/services/Requester', () => ({
    get: jest.fn().mockImplementation((route: string) => new Promise((resolve) => {
        if (route === '/start') {
            resolve({token: '123-123-123'});
        }
    })),
    post: jest.fn().mockReturnValue(Promise.resolve()),
}));

jest.mock('sulu-admin-bundle/stores/ResourceStore', () => jest.fn(function(resourceKey, id) {
    this.resourceKey = resourceKey;
    this.id = id;
}));

jest.mock('sulu-admin-bundle/services/Router', () => jest.fn(function(history) {
    this.history = history;
    this.attributes = {locale: 'de'};
}));

beforeEach(() => {
    jest.resetModules();

    previewConfigStore.setConfig({
        mode: 'auto',
        debounceDelay: 100,
        routes: {},
    });
});

test('Render correct preview', () => {
    const resourceStore = new ResourceStore('pages', 1, {title: 'Test'});
    const router = new Router({});

    expect(shallow(<Preview resourceStore={resourceStore} router={router} />)).toMatchSnapshot();
});

test('Render button to start preview', () => {
    const resourceStore = new ResourceStore('pages', 1, {title: 'Test'});
    const router = new Router({});

    previewConfigStore.setConfig({
        mode: 'on_request',
        debounceDelay: 100,
        routes: {},
    });

    expect(render(<Preview resourceStore={resourceStore} router={router} />)).toMatchSnapshot();
});

test('React and update preview when data is changed', () => {
    const resourceStore = new ResourceStore('pages', 1);
    resourceStore.data = observable.map({title: 'Test'});
    resourceStore.loading = false;
    const router = new Router({});
    const requestPromise = Promise.resolve({token: '123-123-123'});

    Requester.get.mockReturnValue(requestPromise);

    mount(<Preview resourceStore={resourceStore} router={router} />);

    expect(Requester.get).toBeCalledWith('/start');

    resourceStore.data.set('title', 'New Test');

    return requestPromise.then(() => {
        expect(Requester.post).toBeCalledWith('/update', {data: {title: 'New Test'}});
    });
});
