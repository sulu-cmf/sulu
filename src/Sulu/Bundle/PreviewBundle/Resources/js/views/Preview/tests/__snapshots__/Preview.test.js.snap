// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render button to start preview 1`] = `
<button>
  Start
</button>
`;

exports[`Render correct preview 1`] = `
<div
  className="container"
>
  <div
    className="loaderContainer"
  >
    <Loader
      size={40}
    />
  </div>
  <Toolbar
    skin="dark"
  >
    <Controls
      skin="light"
    >
      <Button
        active={false}
        disabled={false}
        hasOptions={false}
        icon="su-arrow-right"
        onClick={[Function]}
        primary={false}
      />
      <Button
        active={false}
        disabled={false}
        hasOptions={false}
        icon="fa-refresh"
        onClick={[Function]}
        primary={false}
      />
    </Controls>
  </Toolbar>
</div>
`;
